src/streamlink/plugins/trovo.py

```diff
@@ -0,0 +1,167 @@
+
+
+@pluginmatcher(re.compile(
+    r"https?://(?:www\.)?trovo\.live/([^/?&]+)(?:/([^/?&]+))?"
+))
+class Trovo(Plugin):
+    @staticmethod
+    def generate_qid():
+        return ("%010x" % random.getrandbits(40)).upper()
+
+    def get_vod(self, video_id):
+        json = self.session.http.post(
+            f"https://gql.trovo.live/?qid={self.generate_qid()}",
+            json=[{
+                "operationName": "batchGetVodDetailInfo",
+                "variables": {
+                    "params": {
+                        "vids": [video_id],
+                    },
+                },
+                "extensions": {
+                    "persistedQuery": {
+                        "version": 1,
+                        "sha256Hash": "ceae0355d66476e21a1dd8e8af9f68de95b4019da2cda8b177c9a2255dad31d0",
+                    },
+                },
+            }],
+            schema=validate.Schema(
+                validate.parse_json(), [{
+                    "data": {
+                        "batchGetVodDetailInfo": {
+                            "VodDetailInfos": validate.any({
+                                video_id: {
+                                    "streamerInfo": {
+                                        "userName": str,
+                                    },
+                                    "vodInfo": {
+                                        "playInfos": [{
+                                            "desc": validate.all(validate.transform(lambda s: s.lower()), str),
+                                            "playUrl": validate.url(),
+                                        }],
+                                        "vid": str,
+                                        "title": str,
+                                        "categoryName": str,
+                                        "playbackRights": {
+                                            "playbackRightsSetting": str,
+                                            "playbackRights": str,
+                                        },
+                                    },
+                                },
+                            }, {}),
+                        },
+                    },
+                }],
+                validate.get((0, "data", "batchGetVodDetailInfo", "VodDetailInfos"))
+            ),
+        )
+
+        if video_id not in json:
+            log.error("Video not found")
+            return
+        else:
+            json = json[video_id]
+
+        log.debug(json["vodInfo"]["playbackRights"])
+        self.id = json["vodInfo"]["vid"]
+        self.author = json["streamerInfo"]["userName"]
+        self.title = json["vodInfo"]["title"]
+        self.category = json["vodInfo"]["categoryName"]
+
+        for s in json["vodInfo"]["playInfos"]:
+            q = s["desc"]
+            if "(source)" in q:
+                q = f"source_{q.replace('(source)', '')}"
+            yield q, HLSStream(self.session, s["playUrl"])
```

Comment:
The `playUrl` values seem to be `http://` based. Accessing `https://` works, so `update_scheme("https:", s["playUrl"])` should be used here.

Same thing for live streams.

Reply:
I hadn't noticed that, thanks.